File: 剑指offer(Sword for offer)/树的子结构.py
# -*- coding:utf-8 -*-
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# 方法1：常规解法，同C++
class Solution:
    def HasSubtree(self, pRoot1, pRoot2):
        # write code here
        if not pRoot1 or not pRoot2:
            return False
        return self.is_subtree(pRoot1, pRoot2) or self.HasSubtree(pRoot1.left, pRoot2) or self.HasSubtree(pRoot1.right, pRoot2)
      
    def is_subtree(self, A, B):
        if not B:
            return True
        if not A or A.val != B.val:
            return False
        return self.is_subtree(A.left,B.left) and self.is_subtree(A.right, B.right)

# 方法2：神级解法，将树转化为前序遍历字符串！
# 但是经过思考、举例，这个代码是有bug的，因为前序遍历不能代表一棵树的结构。
class Solution:
    def HasSubtree(self, pRoot1, pRoot2):
		# 将树转化为前序遍历字符串返回。（相当于把一棵树压扁为一个字符串）
        def convert(p):
            if p: # 当字符串不为空串时
                return str(p.val) +  convert(p.left) + convert(p.right)
            else:
                return ""
		# 看目标字符串（目标树）是否是该字符串的子串
        return convert(pRoot2) in convert(pRoot1) if pRoot2 else False
        

File: 剑指offer(Sword for offer)/树的子结构.cpp
// 输入两棵二叉树A，B，判断B是不是A的子结构。（ps：我们约定空树不是任意一个树的子结构）

#include<iostream>
using namespace std;

struct TreeNode {
	int val;
	struct TreeNode *left;
	struct TreeNode *right;
	TreeNode(int x) :
			val(x), left(NULL), right(NULL) {
	}
};

// 我的方法：这样写是不能通过牛客网的测试用例的。
// 因为牛客网要求的是t2树是否是t1树的“子结构”，而非是否是t1树的子树！
// 我这样写，求得是t2树是否是t1树的子树
class Solution {
public:
    bool HasSubtree(TreeNode* t1, TreeNode* t2)
    {
		if(!t1 || !t2){return false;}
		if(isSameTree(t1, t2)){return true;}
		bool l = isSameTree(t1->left, t2);
		bool r = isSameTree(t1->right, t2);
		return l || r;
    }
	// 工具函数：判断两颗树是否相等
	// 用递归的方法，判断两颗树是不是相等
	bool isSameTree(TreeNode* tree1, TreeNode* tree2){
		if(!tree1 && !tree2){return true;}
		if(!tree1 || !tree2){return false;}
		if(tree1->val != tree2->val){return false;}
		bool l = isSameTree(tree1->left, tree2->left);
		bool r = isSameTree(tree1->right, tree2->right);
		return l && r;
	}
};

// 方法2：用DFS深度搜索遍历的方法，看t2树是否是t1树的某段子结构。
class Solution {
public:
	// 递归
    bool HasSubtree(TreeNode* t1, TreeNode* t2)
    {
		if(!t1 || !t2){return false;}
		return isPartTree(t1, t2) || HasSubtree(t1->left, t2) || HasSubtree(t1->right, t2);
    }
	// 工具函数：DFS判断一颗树是否是另一棵树的一段子结构
	bool isPartTree(TreeNode* tree1, TreeNode* tree2){
		if(!tree2){return true;} // 当tree2已经遍历完成了（为NULL了），说明tree2的每个结点在tree1中都有对应的结点了，返回true
		if(!tree1){return false;} // 当tree2不为空，而tree1为空，返回false
		if(tree1->val == tree2->val){ // 当tree1、2都不空，其元素相等时，递归
			return isPartTree(tree1->left, tree2->left) && isPartTree(tree1->right, tree2->right);
		}else{return false;} // 其元素不相等时，返回false
	}
};

